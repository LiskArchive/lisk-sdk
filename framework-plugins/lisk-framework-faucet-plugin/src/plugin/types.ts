/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export interface FaucetPluginConfig {
	port: number;
	host: string;
	encryptedPrivateKey: string;
	tokenID: string;
	captchaSitekey: string;
	captchaSecretkey: string;
	applicationUrl: string;
	fee: string;
	amount: string;
	tokenPrefix: string;
	captchaSecret: string;
	logoURL?: string;
}

export interface State {
	publicKey?: Buffer;
	privateKey?: Buffer;
	address?: string;
}
