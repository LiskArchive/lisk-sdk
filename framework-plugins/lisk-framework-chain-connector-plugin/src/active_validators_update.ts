/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-bitwise */
import { ActiveValidator, utils, ActiveValidatorsUpdate } from 'lisk-sdk';
import { ValidatorsData } from './types';

/**
 * @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0053.md#computing-the-validators-update
 */

export const calculateActiveValidatorsUpdate = (
	validatorsDataAtLastCertificate: ValidatorsData,
	validatorsDataAtNewCertificate: ValidatorsData,
): { activeValidatorsUpdate: ActiveValidatorsUpdate; certificateThreshold: bigint } => {
	let certificateThreshold;
	// If the certificate threshold is not changed from last certificate then we assign zero
	if (
		validatorsDataAtNewCertificate.certificateThreshold ===
		validatorsDataAtLastCertificate.certificateThreshold
	) {
		certificateThreshold = validatorsDataAtLastCertificate.certificateThreshold;
	} else {
		certificateThreshold = validatorsDataAtNewCertificate.certificateThreshold;
	}

	const activeValidatorsUpdate = getActiveValidatorsUpdate(
		validatorsDataAtLastCertificate.validators,
		validatorsDataAtNewCertificate.validators,
	);

	return { activeValidatorsUpdate, certificateThreshold };
};

/**
 * @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0053.md#computing-the-validators-update
 */
export const getActiveValidatorsUpdate = (
	currentValidators: ActiveValidator[],
	newValidators: ActiveValidator[],
): ActiveValidatorsUpdate => {
	const currentValidatorsMap = new utils.dataStructures.BufferMap<ActiveValidator>();
	const allBLSKeysSet = new utils.dataStructures.BufferSet();
	for (const v of currentValidators) {
		currentValidatorsMap.set(v.blsKey, v);
		allBLSKeysSet.add(v.blsKey);
	}

	const validatorsUpdate: ActiveValidator[] = [];
	const blsKeysUpdate: Buffer[] = [];
	const newBLSKeys = new utils.dataStructures.BufferSet();
	for (const v of newValidators) {
		newBLSKeys.add(v.blsKey);
		allBLSKeysSet.add(v.blsKey);
		const currentValidator = currentValidatorsMap.get(v.blsKey);
		// if new validator does not exist, add to the blsKeysUpdate and validatorsUpdate
		if (!currentValidator) {
			blsKeysUpdate.push(v.blsKey);
			validatorsUpdate.push(v);
			continue;
		}
		// if currentValidator exist, but bft weight is different, then add to the validatorsUpdate
		if (currentValidator.bftWeight !== v.bftWeight) {
			validatorsUpdate.push(v);
		}
	}

	for (const v of currentValidators) {
		// We add the validator to the validatorsUpdate (with 0 BFT weight) if the BLS key is not in the newValidators.
		if (!newBLSKeys.has(v.blsKey)) {
			validatorsUpdate.push({
				blsKey: v.blsKey,
				bftWeight: BigInt(0),
			});
		}
	}
	blsKeysUpdate.sort((a, b) => a.compare(b));
	validatorsUpdate.sort((a, b) => a.blsKey.compare(b.blsKey));

	const bftWeightsUpdate = validatorsUpdate.map(v => v.bftWeight);
	const allBLSKeys = [...allBLSKeysSet];
	allBLSKeys.sort((a, b) => a.compare(b));

	let bitmap = BigInt(0);
	const bitmapSize = Number(BigInt(allBLSKeys.length + 7) / BigInt(8));
	for (const validator of validatorsUpdate) {
		const index = allBLSKeys.findIndex(k => k.equals(validator.blsKey));
		bitmap |= BigInt(1) << BigInt(index);
	}
	const hexBitmap = bitmap.toString(16);
	// prepend zero if length is odd number
	let bftWeightsUpdateBitmap = Buffer.from(
		hexBitmap.length % 2 === 0 ? hexBitmap : `0${hexBitmap}`,
		'hex',
	);
	// if the result does not have enough length, pad zero to the left
	if (bftWeightsUpdateBitmap.length < bitmapSize) {
		bftWeightsUpdateBitmap = Buffer.concat([
			Buffer.alloc(bitmapSize - bftWeightsUpdateBitmap.length),
			bftWeightsUpdateBitmap,
		]);
	}

	return {
		blsKeysUpdate,
		bftWeightsUpdate,
		bftWeightsUpdateBitmap,
	};
};
