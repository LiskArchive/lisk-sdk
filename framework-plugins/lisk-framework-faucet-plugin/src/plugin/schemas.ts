/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const configSchema = {
	$id: '#/plugins/liskFaucet/config',
	type: 'object',
	properties: {
		port: {
			type: 'integer',
			minimum: 1,
			maximum: 65535,
		},
		host: {
			type: 'string',
			format: 'ip',
		},
		encryptedPrivateKey: {
			type: 'string',
			description: 'Encrypted private key of the faucet account',
		},
		tokenID: {
			type: 'string',
			format: 'hex',
			description: 'TokenID of faucet',
		},
		applicationUrl: {
			type: 'string',
			format: 'uri',
			description: 'URL to connect',
		},
		fee: {
			type: 'string',
			description: 'The transaction fee used to faucet an account',
		},
		amount: {
			type: 'string',
			description: 'Number of tokens to fund an account per request',
		},
		tokenPrefix: {
			type: 'string',
			description: 'The token prefix associated with your application',
		},
		logoURL: {
			type: 'string',
			format: 'uri',
			description: 'The URL of the logo used on the UI',
		},
		captchaSecretkey: {
			type: 'string',
			description: 'The re-captcha secret key',
		},
		captchaSitekey: {
			type: 'string',
			description: 'The re-captcha site key',
		},
	},
	required: ['tokenID', 'encryptedPrivateKey', 'captchaSecretkey', 'captchaSitekey'],
	default: {
		port: 4004,
		host: '127.0.0.1',
		applicationUrl: 'ws://localhost:7887/rpc-ws',
		fee: '0.1',
		amount: '100',
		tokenPrefix: 'lsk',
	},
};

export const fundParamsSchema = {
	$id: '/lisk/faucet/fund',
	type: 'object',
	required: ['address'],
	properties: {
		address: {
			type: 'string',
			format: 'hex',
		},
		token: {
			type: 'string',
		},
	},
};

export const authorizeParamsSchema = {
	$id: '/lisk/faucet/auth',
	type: 'object',
	required: ['password', 'enable'],
	properties: {
		password: {
			type: 'string',
		},
		enable: {
			type: 'boolean',
		},
	},
};
