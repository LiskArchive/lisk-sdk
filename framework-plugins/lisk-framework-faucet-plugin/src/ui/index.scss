@import url('https://fonts.googleapis.com/css2?family=Roboto');
@import url('https://fonts.googleapis.com/css2?family=Poppins');

#root {
	font-family: 'Poppins', sans-serif;
}

html,
body {
	margin: 0;
	padding: 0;
}

* {
	box-sizing: border-box;
}
dl,
dt,
dd,
div,
p {
	margin: 0;
	padding: 0;
	white-space: pre-line;
}
img {
	vertical-align: middle;
}
