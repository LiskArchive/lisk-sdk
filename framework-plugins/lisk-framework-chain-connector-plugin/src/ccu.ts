/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import {
	BFTHeights,
	ActiveValidator,
	codec,
	certificateSchema,
	CrossChainUpdateTransactionParams,
	apiClient,
	LastCertificate,
	ccmSchema,
	tree,
	CCMsg,
	OutboxRootWitness,
	db as liskDB,
	crossChainUpdateTransactionParams,
} from 'lisk-sdk';
import { getNextCertificateFromAggregateCommits } from './certificate_generation';
import { calculateActiveValidatorsUpdate } from './active_validators_update';
import { ChainConnectorStore } from './db';
import { BlockHeader, InboxUpdate, CrossChainMessagesFromEvents } from './types';

export class Ccu {
	private readonly _chainConnectorPluginDB!: liskDB.Database;
	private readonly _sidechainChainConnectorStore: ChainConnectorStore;
	private readonly _sidechainAPIClient!: apiClient.APIClient;
	private readonly _ownChainID!: Buffer;
	private readonly _lastCertificate!: LastCertificate;

	public constructor(
		store: ChainConnectorStore,
		chainApiClient: apiClient.APIClient,
		ownChainID: Buffer,
		certificate: LastCertificate,
	) {
		this._sidechainChainConnectorStore = store;
		this._sidechainAPIClient = chainApiClient;
		this._ownChainID = ownChainID;
		this._lastCertificate = certificate;
	}

	public async calculateCCUParams(): Promise<void> {
		const blockHeaders = await this._sidechainChainConnectorStore.getBlockHeaders();
		const aggregateCommits = await this._sidechainChainConnectorStore.getAggregateCommits();
		const validatorsHashPreimages =
			await this._sidechainChainConnectorStore.getValidatorsHashPreImage();
		const bftHeights = await this._sidechainAPIClient.invoke<BFTHeights>('consensus_getBFTHeights');
		// Calculate certificate
		const certificate = getNextCertificateFromAggregateCommits(
			blockHeaders,
			aggregateCommits,
			validatorsHashPreimages,
			bftHeights,
			this._lastCertificate,
		);
		if (!certificate) {
			throw new Error('No certificate found in the ccu creation process.');
		}
		// Calculate activeValidatorsUpdate
		let activeValidatorsUpdate: ActiveValidator[] = [];
		let certificateThreshold = BigInt(0);
		const blockHeader = blockHeaders.find(header => header.height === certificate.height);
		if (!blockHeader) {
			throw new Error('No block header found for the given certificate height.');
		}
		if (!this._lastCertificate.validatorsHash.equals(certificate.validatorsHash)) {
			const validatorsUpdateResult = calculateActiveValidatorsUpdate(
				blockHeader,
				validatorsHashPreimages,
				this._lastCertificate,
			);
			activeValidatorsUpdate = validatorsUpdateResult.activeValidatorsUpdate;
			certificateThreshold = validatorsUpdateResult.certificateThreshold;
		}
		// Calculate inboxUpdate
		const inboxUpdates = await this._calculateInboxUpdate(blockHeader);

		// Now create CCUs for all the inboxUpdates handling full or partial updates
		for (const inboxUpdate of inboxUpdates) {
			const serializedCertificate = codec.encode(certificateSchema, certificate);
			const serializedCCUParams = codec.encode(crossChainUpdateTransactionParams, {
				sendingChainID: this._ownChainID,
				activeValidatorsUpdate,
				certificate: serializedCertificate,
				certificateThreshold,
				inboxUpdate,
			} as CrossChainUpdateTransactionParams);
			await this._createCCU(serializedCCUParams);
		}
	}

	private async _calculateInboxUpdate(newBlockHeader: BlockHeader): Promise<InboxUpdate[]> {
		const ccmsFromEvents = await this._sidechainChainConnectorStore.getCrossChainMessages();
		const ccmsListOfList = this._getListOfCCMs(ccmsFromEvents, newBlockHeader);

		if (ccmsListOfList.length === 1) {
			const ccms = ccmsListOfList[0];
			const crossChainMessages = ccms.map(ccm => codec.encode(ccmSchema, ccm));
			// Take the inclusion proof of the last ccm height
			const { inclusionProof } = ccms[ccms.length - 1];

			return [
				{
					crossChainMessages,
					messageWitnessHashes: [],
					outboxRootWitness: inclusionProof,
				},
			];
		}

		// Calculate list of inboxUpdates to be sent by multiple CCUs
		const inboxUpdates = [];
		for (const subList of ccmsListOfList) {
			const crossChainMessages = subList.map(ccm => codec.encode(ccmSchema, ccm));
			// Take the inclusion proof of the last ccm height
			const { inclusionProof } = subList[subList.length - 1];

			// Calculate message witnesses

			const merkleTree = new tree.MerkleTree({ db: this._chainConnectorPluginDB });
			for (const ccm of crossChainMessages) {
				await merkleTree.append(ccm);
			}
			const messageWitnessHashes = await merkleTree.generateRightWitness(crossChainMessages.length);

			inboxUpdates.push({
				crossChainMessages,
				messageWitnessHashes,
				outboxRootWitness: inclusionProof,
			});
		}

		return inboxUpdates;
	}

	private _getListOfCCMs(
		_crossChainMessages: CrossChainMessagesFromEvents[],
		_newBlockHeader: BlockHeader,
	): {
		ccm: CCMsg;
		height: number;
		inclusionProof: OutboxRootWitness;
	}[][] {
		return [[]];
	}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	private async _createCCU(_serializedCCUParams: Buffer) {}
}
