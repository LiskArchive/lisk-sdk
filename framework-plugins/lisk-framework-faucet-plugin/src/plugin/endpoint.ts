/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import axios from 'axios';
import {
	BasePluginEndpoint,
	PluginEndpointContext,
	validator as liskValidator,
	cryptography,
	transactions,
	BasePlugin,
} from 'lisk-sdk';
import { authorizeParamsSchema, fundParamsSchema } from './schemas';
import { FaucetPluginConfig, State } from './types';

// disabled for type annotation
// eslint-disable-next-line prefer-destructuring
const validator: liskValidator.LiskValidator = liskValidator.validator;

export class Endpoint extends BasePluginEndpoint {
	private _state: State = { publicKey: undefined, passphrase: undefined };
	private _client!: BasePlugin['apiClient'];
	private _config!: FaucetPluginConfig;

	public init(state: State, apiClient: BasePlugin['apiClient'], config: FaucetPluginConfig) {
		this._state = state;
		this._client = apiClient;
		this._config = config;
	}
	// eslint-disable-next-line @typescript-eslint/require-await
	public async authorize(context: PluginEndpointContext): Promise<{ result: string }> {
		validator.validate(authorizeParamsSchema, context.params);

		const { enable, password } = context.params;

		try {
			const parsedEncryptedPassphrase = cryptography.encrypt.parseEncryptedMessage(
				this._config.encryptedPassphrase,
			);

			const passphrase = await cryptography.encrypt.decryptMessageWithPassword(
				parsedEncryptedPassphrase,
				password as string,
				'utf-8',
			);

			const { publicKey } = cryptography.legacy.getPrivateAndPublicKeyFromPassphrase(passphrase);

			this._state.publicKey = enable ? publicKey : undefined;
			this._state.passphrase = enable ? passphrase : undefined;
			const changedState = enable ? 'enabled' : 'disabled';

			return {
				result: `Successfully ${changedState} the faucet.`,
			};
		} catch (error) {
			throw new Error('Password given is not valid.');
		}
	}

	public async fundTokens(context: PluginEndpointContext): Promise<{ result: string }> {
		validator.validate(fundParamsSchema, context.params);
		const { address, token } = context.params;

		if (!this._state.publicKey || !this._state.passphrase) {
			throw new Error('Faucet is not enabled.');
		}

		const captchaResult = await axios({
			method: 'post',
			url: 'https://www.google.com/recaptcha/api/siteverify',
			params: {
				secret: this._config.captchaSecretkey,
				response: token,
			},
		});

		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
		if (!captchaResult?.data?.success) {
			throw new Error('Captcha response was invalid.');
		}

		await this._transferFunds(address as string);

		return {
			result: `Successfully funded account at address: ${address as string}.`,
		};
	}

	private async _transferFunds(address: string): Promise<void> {
		const transferTransactionParams = {
			amount: transactions.convertLSKToBeddows(this._config.amount),
			recipientAddress: address,
			data: '',
		};

		const transaction = await this._client.transaction.create(
			{
				moduleID: cryptography.utils.intToBuffer(2, 4).toString('hex'),
				commandID: cryptography.utils.intToBuffer(0, 4).toString('hex'),
				senderPublicKey: this._state.publicKey?.toString('hex'),
				fee: transactions.convertLSKToBeddows(this._config.fee), // TODO: The static fee should be replaced by fee estimation calculation
				params: transferTransactionParams,
			},
			this._state.passphrase as string,
		);

		await this._client.transaction.send(transaction);
	}
}
