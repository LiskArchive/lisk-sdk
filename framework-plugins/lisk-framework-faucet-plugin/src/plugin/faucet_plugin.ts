/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { BasePlugin, PluginInitContext, cryptography } from 'lisk-sdk';
import * as express from 'express';
import { join } from 'path';
import { Server } from 'http';
import { configSchema } from './schemas';
import { FaucetPluginConfig, State } from './types';
import { Endpoint } from './endpoint';

export class FaucetPlugin extends BasePlugin<FaucetPluginConfig> {
	public configSchema = configSchema;
	public endpoint = new Endpoint();

	private _server!: Server;
	private readonly _state: State = { publicKey: undefined, passphrase: undefined };

	public get nodeModulePath(): string {
		return __filename;
	}

	public async init(context: PluginInitContext): Promise<void> {
		await super.init(context);
		this.endpoint.init(this._state, this.apiClient, this.config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await, class-methods-use-this
	public async load(): Promise<void> {
		const app = express();
		app.get('/api/config', (_req, res) => {
			const config = {
				applicationUrl: this.config.applicationUrl,
				amount: this.config.amount,
				tokenPrefix: this.config.tokenPrefix,
				captchaSitekey: this.config.captchaSitekey,
				logoURL: this.config.logoURL,
				faucetAddress: this._state.publicKey
					? cryptography.address.getLisk32AddressFromPublicKey(this._state.publicKey)
					: undefined,
			};
			res.json(config);
		});
		app.use(express.static(join(__dirname, '../../build')));
		this._server = app.listen(this.config.port, this.config.host);
	}

	public async unload(): Promise<void> {
		return new Promise((resolve, reject) => {
			this._server.close(err => {
				if (err) {
					reject(err);
					return;
				}
				resolve();
			});
		});
	}
}
