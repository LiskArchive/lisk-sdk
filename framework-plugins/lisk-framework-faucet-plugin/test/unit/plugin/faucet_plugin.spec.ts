/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { ApplicationConfigForPlugin, testing } from 'lisk-sdk';
import { FaucetPlugin } from '../../../src/plugin';

const appConfigForPlugin: ApplicationConfigForPlugin = {
	...testing.fixtures.defaultConfig,
};
const logger = testing.mocks.loggerMock;

const validOptions = {
	captchaSitekey: '123',
	captchaSecretkey: '123',
	encryptedPrivateKey:
		'kdf=argon2id&cipher=aes-256-gcm&version=1&ciphertext=99053afb5eb03999f06201a0099912cec7826cdc8d9f41fc64d575670128a532f922961353253b4e73815890ed210f637a98e5084e4521a1495f0b793184c5385a965fe70837a8a638f0667c8f803d84872274c4ddc046591c03bed5d141b2193358da2b8db4548ff2dd137dbc796b6fb29acc53362c3a1ea8dd212270a9e2c1&mac=909eba6a8b1b42812f3de13fcdcbe124acd52a40b7ca7e3e9179a2cb8bffa452&salt=082b9d532290492caa842b804517def0&iv=a2947e6a05031e5b4d250604&tag=ca7dab52dccfa9cdb6a74906e84ed625&iterations=1&parallelism=4&memorySize=2024',
};

describe('FaucetPlugin', () => {
	describe('configSchema', () => {
		it('should return valid config schema with default options', () => {
			const plugin = new FaucetPlugin();

			expect(plugin.configSchema).toMatchSnapshot();
		});
	});

	describe('init', () => {
		it('should load default config', async () => {
			const plugin = new FaucetPlugin();
			plugin['_apiClient'] = {
				schema: {},
				invoke: jest.fn(),
			};
			await plugin.init({
				config: validOptions,
				appConfig: appConfigForPlugin,
				logger,
			});

			expect(plugin.config).toMatchSnapshot();
		});

		it('should throw error if valid options are not passed', async () => {
			const plugin = new FaucetPlugin();
			plugin['_apiClient'] = {
				schema: {},
				invoke: jest.fn(),
			};
			await expect(
				plugin.init({
					config: { captchaSitekey: '123', captchaSecretkey: '123' },
					appConfig: appConfigForPlugin,
					logger,
				}),
			).rejects.toThrow(
				"Lisk validator found 1 error[s]:\nMissing property, must have required property 'encryptedPrivateKey'",
			);
		});

		it('should load custom tokenPrefix', async () => {
			const plugin = new FaucetPlugin();
			plugin['_apiClient'] = {
				schema: {},
				invoke: jest.fn(),
			};
			await plugin.init({
				config: { ...validOptions, tokenPrefix: 'myToken' },
				appConfig: appConfigForPlugin,
				logger,
			});

			expect(plugin.config.tokenPrefix).toBe('myToken');
		});

		it('should load custom config values', async () => {
			const options = {
				...validOptions,
				tokenPrefix: 'myToken2',
				fee: '100000000',
				token: '1111111111111',
				captcha: { config: { value: 'value' } },
			} as never;
			const plugin = new FaucetPlugin();
			plugin['_apiClient'] = {
				schema: {},
				invoke: jest.fn(),
			};
			await plugin.init({
				config: options,
				appConfig: appConfigForPlugin,
				logger,
			});

			expect(plugin.config).toMatchObject(expect.objectContaining(options));
		});
	});
});
