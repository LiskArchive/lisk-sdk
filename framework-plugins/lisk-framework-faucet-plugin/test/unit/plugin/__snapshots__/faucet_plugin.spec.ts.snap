// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaucetPlugin configSchema should return valid config schema with default options 1`] = `
{
  "$id": "#/plugins/liskFaucet/config",
  "default": {
    "amount": "100",
    "applicationUrl": "ws://localhost:8080/ws",
    "fee": "0.1",
    "host": "127.0.0.1",
    "port": 4004,
    "tokenPrefix": "lsk",
  },
  "properties": {
    "amount": {
      "description": "Number of tokens to fund an account per request",
      "type": "string",
    },
    "applicationUrl": {
      "description": "URL to connect",
      "format": "uri",
      "type": "string",
    },
    "captchaSecretkey": {
      "description": "The re-captcha secret key",
      "type": "string",
    },
    "captchaSitekey": {
      "description": "The re-captcha site key",
      "type": "string",
    },
    "encryptedPassphrase": {
      "description": "Encrypted passphrase of the genesis account",
      "format": "encryptedPassphrase",
      "type": "string",
    },
    "fee": {
      "description": "The transaction fee used to faucet an account",
      "type": "string",
    },
    "host": {
      "format": "ip",
      "type": "string",
    },
    "logoURL": {
      "description": "The URL of the logo used on the UI",
      "format": "uri",
      "type": "string",
    },
    "port": {
      "maximum": 65535,
      "minimum": 1,
      "type": "integer",
    },
    "tokenPrefix": {
      "description": "The token prefix associated with your application",
      "type": "string",
    },
  },
  "required": [
    "encryptedPassphrase",
    "captchaSecretkey",
    "captchaSitekey",
  ],
  "type": "object",
}
`;

exports[`FaucetPlugin init should load default config 1`] = `
{
  "amount": "100",
  "applicationUrl": "ws://localhost:8080/ws",
  "captchaSecretkey": "123",
  "captchaSitekey": "123",
  "encryptedPassphrase": "salt=683425ca06c9ff88a5ab292bb5066dc5&cipherText=4ce151&iv=bfaeef79a466e370e210f3c6&tag=e84bf097b1ec5ae428dd7ed3b4cce522&version=1",
  "fee": "0.1",
  "host": "127.0.0.1",
  "port": 4004,
  "tokenPrefix": "lsk",
}
`;
